app/models/physical_server.rb

```diff
@@ -104,4 +104,24 @@ def v_availability
   def v_host_os
     host.try(:vmm_product).nil? ? N_("") : host.vmm_product
   end
+
+  #
+  # Searches the switches that are connected to the physical server
+  #   it evaluates the server physical network ports that are
+  #   connected with switch physical network ports
+  #
+  # @returns [Array<PhysicalSwitch>] - a list of connected switches,
+  #   and if it isn't connected to any, returns an empty list.
+  #
+  def switches
+    switches = []
+
+    hardware.nics&.each do |network_device|
+      network_device.physical_network_ports&.each do |port|
+        switches << PhysicalSwitch.find(port.connected_port.switch_id) if port.connected_port.try(:switch_id).present?
```

Comment:
You should do this through rails associations instead of loops and finds.

Give this a shot:
```
diff --git a/app/models/guest_device.rb b/app/models/guest_device.rb
index cd4791532a..09db2943cc 100644
--- a/app/models/guest_device.rb
+++ b/app/models/guest_device.rb
@@ -16,6 +16,7 @@ class GuestDevice < ApplicationRecord
   has_many :child_devices, -> { where(:parent_device_id => ids) }, :foreign_key => "parent_device_id", :class_name => "GuestDevice", :dependent => :destroy
 
   has_many :physical_network_ports, :dependent => :destroy
+  has_many :connected_physical_switches, :through => :physical_network_ports
 
   alias_attribute :name, :device_name
 
diff --git a/app/models/hardware.rb b/app/models/hardware.rb
index af0d5d155e..4168d913ba 100644
--- a/app/models/hardware.rb
+++ b/app/models/hardware.rb
@@ -22,6 +22,7 @@ class Hardware < ApplicationRecord
   has_many    :nics, -> { where("device_type = 'ethernet'") }, :class_name => "GuestDevice", :foreign_key => :hardware_id
   has_many    :ports, -> { where("device_type != 'storage'") }, :class_name => "GuestDevice", :foreign_key => :hardware_id
   has_many    :physical_ports, -> { where("device_type = 'physical_port'") }, :class_name => "GuestDevice", :foreign_key => :hardware_id
+  has_many    :connected_physical_switches, :through => :guest_devices
 
   virtual_column :ipaddresses,   :type => :string_set, :uses => :networks
   virtual_column :hostnames,     :type => :string_set, :uses => :networks
diff --git a/app/models/physical_network_port.rb b/app/models/physical_network_port.rb
index 3b43db5b82..cb627d900f 100644
--- a/app/models/physical_network_port.rb
+++ b/app/models/physical_network_port.rb
@@ -1,6 +1,7 @@
 class PhysicalNetworkPort < ApplicationRecord
   belongs_to :guest_device
-  belongs_to :physical_switch
+  belongs_to :physical_switch, :foreign_key => :switch_id
 
   has_one :connected_port, :foreign_key => "connected_port_uid", :primary_key => "uid_ems", :class_name => "PhysicalNetworkPort", :dependent => :nullify
+  has_one :connected_physical_switch, :through => :connected_port, :source => :physical_switch

diff --git a/app/models/physical_server.rb b/app/models/physical_server.rb
index 275637078d..1d598fce2c 100644
--- a/app/models/physical_server.rb
+++ b/app/models/physical_server.rb
@@ -106,23 +106,5 @@ class PhysicalServer < ApplicationRecord
     host.try(:vmm_product).nil? ? N_("") : host.vmm_product
   end
 
+  has_many :physical_switches, :through => :hardware, :source => :connected_physical_switches
```

This way I was able to do:
```
>> server.physical_switches
  PhysicalSwitch Load (1.0ms)  SELECT "switches".* FROM "switches" INNER JOIN "physical_network_ports" ON "switches"."id" = "physical_network_ports"."switch_id" INNER JOIN "physical_network_ports" "physical_network_ports_physical_switches" ON "physical_network_ports"."connected_port_uid" = "physical_network_ports_physical_switches"."uid_ems" INNER JOIN "guest_devices" ON "physical_network_ports_physical_switches"."guest_device_id" = "guest_devices"."id" INNER JOIN "hardwares" ON "guest_devices"."hardware_id" = "hardwares"."id" INNER JOIN "computer_systems" ON "hardwares"."computer_system_id" = "computer_systems"."id" WHERE "switches"."type" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch') AND "computer_systems"."managed_entity_id" = $1 AND "computer_systems"."managed_entity_type" = $2  [["managed_entity_id", 1], ["managed_entity_type", "PhysicalServer"]]
  PhysicalSwitch Inst Including Associations (16.0ms - 1rows)
=> #<ActiveRecord::Associations::CollectionProxy [#<PhysicalSwitch id: 1, name: nil, ports: nil, created_on: "2018-06-22 07:35:54", updated_on: "2018-06-22 07:35:54", uid_ems: nil, allow_promiscuous: nil, forged_transmits: nil, mac_changes: nil, switch_uuid: nil, shared: nil, mtu: nil, ems_id: nil, type: "PhysicalSwitch", health_state: nil, power_state: nil>]>
```

Reply:
These changes seems to works fine, but, if I try to call (which is what the topology page do)
```ruby
PhysicalServer.includes(:physical_switches)
```
I get the following error:
```ruby
PhysicalServer Load (0.8ms)  SELECT "physical_servers".* FROM "physical_servers"
  PhysicalServer Inst Including Associations (587.0ms - 3rows)
  ComputerSystem Load (0.5ms)  SELECT "computer_systems".* FROM "computer_systems" WHERE "computer_systems"."managed_entity_type" = $1 AND "computer_systems"."managed_entity_id" IN (1, 2, 3)  [["managed_entity_type", "PhysicalServer"]]
  ComputerSystem Inst Including Associations (3.5ms - 3rows)
  Hardware Load (0.6ms)  SELECT "hardwares".* FROM "hardwares" WHERE "hardwares"."computer_system_id" IN (1, 2, 3)
  Hardware Inst Including Associations (16.4ms - 3rows)
  GuestDevice Load (0.4ms)  SELECT "guest_devices".* FROM "guest_devices" WHERE "switches"."type" IN ('PhysicalSwitch', 'ManageIQ::Providers::Lenovo::PhysicalInfraManager::PhysicalSwitch') AND "guest_devices"."hardware_id" IN (1, 2, 3)
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "switches"
LINE 1: ...LECT "guest_devices".* FROM "guest_devices" WHERE "switches"...
```